libraries/fc/src/log/logger_config.cpp

```diff
@@ -49,6 +62,12 @@ namespace fc {
          iter.second->initialize( ios );
    }
 
+   void log_config::shutdown_appenders() {
+      std::lock_guard g( log_config::get().log_mutex );
+      for( auto& iter : log_config::get().appender_map )
+         iter.second->shutdown();
```

Comment:
Indirect calls while holding a mutex are dangerous.  In this case it can deadlock:
- this thread blocks joining the zipkin thread
- the zipkin thread blocks in dlog or wlog trying to acquire the `log_mutex`

Reply:
I agree this is a bad pattern. However, it follows initialize and log which are also called holding the lock. I think I can just reset the lock guard in `zipkin_appender::shutdown()`

Reply:
Actually, I forgot why this was all even needed. It is needed because the loggers explicitly leak.
https://github.com/EOSIO/fc/blob/master/src/log/logger_config.cpp#L16
So I wanted a hook that would allow for purging of all queued up zipkin messages.
This is not very clean. :(

Reply:
The thread has to be joined, or you have undefined behavior (unless it only executes functions that are permitted in a signal handler, which I greatly doubt.).

Reply:
BTW, this same deadlock is possible on `configure_logging` where appenders are destroyed via sighup.